### PR TITLE
Merge axlsx_styler gem into caxlsx

### PR DESCRIPTION
**Pull request by westonganger**
### Description
Merge axlsx_styler gem into caxlsx. As brought up in #167.

### Your checklist for this pull request

- [ ] I have reviewed the [guidelines for contributing](../blob/master/CONTRIBUTING.md) to this repository.
- [ ] I have added (or updated) appropriate tests if this PR fixes a bug or adds a feature.
- [ ] If this PR doesn't need tests (docs change), I added `[ci skip]` to the title of the PR.
- [ ] If this closes any issues, I have added "Closes `#issue`" to the PR description or my commit messages.
- [ ] I have updated the documentation accordingly.
- [ ] All new and existing tests passed.
- [ ] I added an entry to the [changelog](../blob/master/CHANGELOG.md).

### After this is merged

We need to make some changes to `axlsx_styler` some short time after a successful version release of caxlsx

1. Add a notice on the top of the axlsx_styler readme
2. Cut a new release which raises an exception if a newer non-supported caxlsx version is used

### CODE REVIEW

**Comment by straydogstudio** [comment]
Thanks for the quick work @westonganger

File: lib/axlsx.rb [hunk 1]
```diff
@@ -166,6 +166,19 @@ def self.booleanize(value)
     end
   end
 
+  # utility method for performing a deep merge on a Hash
+  # @param [Hash] Hash to merge into
+  # @param [Hash] Hash to be added
+  def self.hash_deep_merge(first_hash, second_hash)
```
Comment (westonganger) [inline lib/axlsx.rb:172]: Modeled after https://github.com/rails/rails/blob/main/activesupport/lib/active_support/core_ext/hash/deep_merge.rb

File: lib/axlsx/workbook/workbook.rb [hunk 2]
```diff
@@ -187,6 +187,35 @@ def styles
       @styles
     end
 
+    # An array that holds all cells with styles
+    def styled_cells
+      @styled_cells ||= Set.new
+    end
+
+    # Checks if styles are indexed to make it work for pre 0.1.5 version
```
Comment (kiskoza) [inline lib/axlsx/workbook/workbook.rb:195]: I don't think we still need this check, we should not bring over legacy codes

File: examples/add_border_example.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  sheet.add_border "B3:D3", { edges: [:top], style: :thick }
+  sheet.add_border ["C3:C4", "D3:D4"]
+end
+
+p.serialize "add_border.xlsx"
+```
+
+## Output
+
+![Output](images/add_border.png "Output")
```
Comment (kiskoza) [inline examples/add_border_example.md:28]: the screenshot seems to be missing, also we should name it the same as the .md file (e.g.: add_broder_example.png)

**Review by kiskoza — CHANGES_REQUESTED** [review]
Thanks for the PR, I left a few comments we need to fix. I saw a few TODOs in the comments, can we address them? (e.g: remove them if they are not relevant). Also there is a test/tc_axlsx file which is empy, maybe by an accident?

File: lib/axlsx/workbook/workbook.rb (hunk 2 above)
Comment (westonganger) [reply lib/axlsx/workbook/workbook.rb]: Ok I updated the comment. However this is actually a useful method/attr

File: examples/add_border_example.md (hunk 3 above)
Comment (westonganger) [reply examples/add_border_example.md:28]: Added the screenshots

**Comment by westonganger** [comment]
@kiskoza ok I have resolved all the TODO's and addressed your comments.

**Comment by westonganger** [comment]
Seems to me this is more or less ready to go. Any more comments here.

**Review by kiskoza — APPROVED** [review]
Sorry for the delay, it was a busy week for me. Thanks for solving all my comments, it looks good to me, ready to merge :+1:

**Comment by straydogstudio** [comment]
Do we want to release a 3.2.1 version?

**Comment by westonganger** [comment]
I have now released v1.2.0 of axlsx_styler which drops support for 3.3.0+ and added a notice at the top of the axlsx_styler readme

**Pull request merged**